### DISCUSSION TITLE
@的黑魔法：Python 裝飾器 (decorator) 的魔力🌠

### DISCUSSION THREAD

**Discussion by RogelioKG** [1101-GS4719-C/Python-Community-GS4719-C#1506] [General]
[**🌠 筆記：閉包與裝飾器的黑魔法**](https://www.notion.so/rogeliokg/decorator-4e153d0178a4401a9d0f3d125ea385b4?pvs=4)

+ 本篇筆記是採用 Notion 紀錄，如果你不是很熟悉 Notion 的介面操作，可以參考[這支影片](https://youtu.be/Q_PfYlAtvHc?si=kzKqN8T-ilA1dcG6)。
+ 如果你想在 Notion 開啟黑暗模式，快捷鍵是 ctrl + shift + L (Windows) / cmd + shift + L (Mac)，我的筆記推薦使用黑暗模式瀏覽。
+ 如果你覺得某處的內容有問題，可以直接在該處內容處留下 comment 讓我看見，我應該是有開啟 comment 功能的。
+ 如果你看到網頁內容正在動態改變，請不要害怕，一定是因為我又看到了某個酷酷的裝飾器所以開始筆記🥹。

**Comment by brian431** [comment]
謝謝您整理的筆記，若是要了解Python，對裝飾器是必須要有一定程度的了解的。因為他的應用十分廣泛且經常在各套件的程式碼中出現。

**Comment by pudding0803** [comment]
好神奇😮，在看到你這篇文章之前，我都不知道 @ 那行的意思是什麼😅，現在終於了解了，感謝分享！

不然我目前其實也都只有在 Flask 之類的框架看到過

**Reply by RogelioKG** [reply]
你才是大佬，我好菜QQ

**Comment by sumairu0727** [comment]
感謝大大整理，但如果我需要還是會撬開你的筆電讓你教我的<3

**Reply by RogelioKG** [reply]
超呵🥴

**Comment by lin157644** [comment]
感謝您鉅細靡遺的介紹，不管是用於學習或是複習都是很優質的文章，我自己過去因為對於Closure的概念沒有那麼熟悉因此比較少使用裝飾器，在經過編譯器課程的荼毒後，發現Python裝飾器這種Closure的運用形式與Syntax sugar十分的便利，也嘗試在近期的專案中引入裝飾器的用法。

**Comment by 1chooo** [comment]
看完這篇文章後覺得非常有幫助，未來有需求會再回訪的！非常感謝！

**Comment by RogelioKG** [comment]
各位大大早安。
下學期開始，家凱老師應該會開始上 Python 入門，到時候這篇文章或許會有非常多人參考。
由於之前的文章有太多疏漏，小弟決定將本篇替換成個人的 Notion 的裝飾器筆記。
裡面會從閉包開始講解，一步一步用觀察式的手法拆解裝飾器的魔法。
下方也有裝飾器的資料庫，會把我所找到的一些有趣且實用的裝飾器一併記錄下來。
(小弟在去年很幸運的成功轉入 CSIE 🥹，終於達成我的夢想，希望未來能不斷變得更強，加油，奧利給！)
